Library/OcMainLib/OpenCoreUefiUnloadDrivers.c

```diff
@@ -0,0 +1,352 @@
+#include <Library/DevicePathLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/PrintLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+
+#include <Library/OcConfigurationLib.h>
+
+#include <PiDxe.h>
+
+#include <Protocol/ComponentName2.h>
+#include <Protocol/ComponentName.h>
+#include <Protocol/DriverBinding.h>
+#include <Protocol/FirmwareVolume2.h>
+#include <Protocol/LoadedImage.h>
+
+typedef
+VOID
+(*PROCESS_IMAGE)(
+  IN VOID          *Context,
+  IN CONST CHAR16  *Name,
+  IN EFI_HANDLE    Handle
+  );
+
+typedef struct {
+  CHAR16     *UnloadName;
+  BOOLEAN    Unloaded;
+} UNLOAD_INFO;
+
+typedef struct {
+  UINTN          UnloadNameCount;
+  UNLOAD_INFO    *UnloadInfo;
+} UNLOAD_IMAGE_CONTEXT;
+
+typedef struct {
+  CHAR8    *FileBuffer;
+  UINTN    FileBufferSize;
+} DRIVER_REPORT_CONTEXT;
+
+/**
+  Get the best supported language for this driver.
+
+  First base on the user input language  to search, otherwise get the first language
+  from the supported language list. The caller needs to free the best language buffer.
+
+  @param[in] SupportedLanguages      The supported languages for this driver.
+  @param[in] InputLanguage           The user input language.
+  @param[in] Iso639Language          Whether get language for ISO639.
+
+  @return                            The best supported language for this driver.
+**/
+CHAR8 *
+EFIAPI
+GetBestLanguageForDriver (
+  IN CONST CHAR8  *SupportedLanguages,
+  IN CONST CHAR8  *InputLanguage,
+  IN BOOLEAN      Iso639Language
+  )
+{
+  CHAR8  *BestLanguage;
+
+  BestLanguage = GetBestLanguage (
+                   SupportedLanguages,
+                   Iso639Language,
+                   (InputLanguage != NULL) ? InputLanguage : "",
+                   SupportedLanguages,
+                   NULL
+                   );
+
+  return BestLanguage;
+}
+
+/**
+  Function to retrieve the driver name (if possible) from the ComponentName or
+  ComponentName2 protocol
+
+  @param[in] TheHandle      The driver handle to get the name of.
+  @param[in] Language       The language to use.
+
+  @retval NULL              The name could not be found.
+  @return                   A pointer to the string name.  Do not de-allocate the memory.
+**/
+CONST CHAR16 *
+EFIAPI
+GetStringNameFromHandle (
+  IN CONST EFI_HANDLE  TheHandle,
+  IN CONST CHAR8       *Language
+  )
+{
+  EFI_COMPONENT_NAME2_PROTOCOL  *CompNameStruct;
+  EFI_STATUS                    Status;
+  CHAR16                        *RetVal;
+  CHAR8                         *BestLang;
+
+  BestLang = NULL;
+
+  Status = gBS->OpenProtocol (
+                  TheHandle,
+                  &gEfiComponentName2ProtocolGuid,
+                  (VOID **)&CompNameStruct,
+                  gImageHandle,
+                  NULL,
+                  EFI_OPEN_PROTOCOL_GET_PROTOCOL
+                  );
+  if (!EFI_ERROR (Status)) {
+    BestLang = GetBestLanguageForDriver (CompNameStruct->SupportedLanguages, Language, FALSE);
```

Comment:
Hmm, why not always pass NULL?

Reply:
Sorry, for what? Language? If so, in fact we do - but GetStringNameFromHandle is from UefiHandleParsingLib.c, and takes Language - to which I always pass NULL.

Reply:
No, for BestLang. Clearly, when the callee returns NULL , it is passed to GetDriverName (so it’s allowed? On the phone, cannot check). I have two concerns: 1) Reducing calls to anything HII where-ever possible and 2) keeping the keys invariant of the system configuration, i.e. if the UEFI language is changed, the driver name keys used in OC should not suddenly become invalid (because the “BestLang” is suddenly French instead of English).

Reply:
BestLang definitely cannot be NULL: https://github.com/acidanthera/audk/blob/master/MdePkg/Include/Protocol/ComponentName.h#L40 https://github.com/acidanthera/audk/blob/master/MdePkg/Include/Protocol/ComponentName2.h#L49

GetBestLanguageForDriver, as called, basically chooses the first language which _the driver itself_ lists as available (this list never changes, for a given driver). So this call to GetBestLanguageForDriver is already intended (i.e. by me) not to change if the firmware language is changed (in fact, precisely by me not passing in any specified/desired language in the Language parameter).

Reply:
If it cannot be NULL, then why call the function when it is NULL? That's where I drew that conclusion from.

Right, the callee does not retrieve PlatformLang, the caller does - good. I do wonder about the empty string though, maybe we should default to US English in case there are bugs?

Reply:
Well, as I said, in case there are bugs - e.g. SupportedLanguages is empty or otherwise borked. But can just wait whether there are any such reports, I was just thinking aloud.

Reply:
Also, I didn't write any of this, including the empty string thing. Also it's not calling a protocol even then, it's calling linked MdePkg UefiLib code, so I guess we know there isn't a bug?

Reply:
> Well, as I said, in case there are bugs - e.g. SupportedLanguages is empty or otherwise borked. But can just wait whether there are any such reports, I was just thinking aloud.

Sounds like you're more or less okay with this PR then, I think, once I've added something to the docs mentioning that one reason to avoid unloading images, normally, is borked drivers which crash when you try to unload them?